File: Problem_1.py
# Time Complexity: O(n) for __init__, O(1) for next, O(1) for hasNext
# Space Complexity: O(n)
# Did this code successfully run on Leetcode: Yes
# Any problem you faced while coding this: No

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator(object):
    def __init__(self, root):
        """
        :type root: TreeNode
        """
        self.stack = []
        self.pushUntilNone(root)

    def next(self):
        """
        :rtype: int
        """
        root = self.stack.pop()
        self.pushUntilNone(root.right)
        return root.val

    def hasNext(self):
        """
        :rtype: bool
        """
        return self.stack

    def pushUntilNone(self, root):
        while root:
            self.stack.append(root)
            root = root.left

# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()

File: problem1.go
// Time Complexity : O(1)
// Space Complexity : O(H) - Where H is height of the tree
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
//TC - O(1)
//SC - O(H)
//var stack []int
type BSTIterator struct {
    stack []*TreeNode
}


func Constructor(root *TreeNode) BSTIterator {
    iterator := BSTIterator{
        stack: []*TreeNode{},
    }
    
    dfs(root,&iterator)
    
    return iterator
    
}


func (this *BSTIterator) Next() int {
    
    next := this.stack[len(this.stack)-1]
    this.stack = this.stack[:len(this.stack)-1]
    if next.Right != nil{
        dfs(next.Right,this)
    }
    
    return next.Val
}


func (this *BSTIterator) HasNext() bool {
    if len(this.stack) > 0{
        return true
    }
    
    return false
}

func dfs(root *TreeNode,iterator *BSTIterator){
    for root != nil{
        iterator.stack = append(iterator.stack,root)
        root = root.Left
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * obj := Constructor(root);
 * param_1 := obj.Next();
 * param_2 := obj.HasNext();
 */
